extra/src/main/java/aima/extra/probability/BigDecimalProbabilityNumber.java

```diff
@@ -0,0 +1,381 @@
+package aima.extra.probability;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.MathContext;
+import java.math.RoundingMode;
+
+/**
+ * Implementation based on the underlying BigDecimal datatype. Supports higher
+ * precision than DoubleProbabilityNumber (DECIMAL128 precision by default, but
+ * can go upto UNLIMITED). The class is immutable.
+ * 
+ * @author Nagaraj Poti
+ *
+ */
+public class BigDecimalProbabilityNumber implements ProbabilityNumber, Comparable<ProbabilityNumber> {
+
+	// Constants
+
+	/**
+	 * IEEE 754R Decimal128 format - Precision of 34 digits and a rounding mode
+	 * of HALF_EVEN. HALF_EVEN rounding mode statistically minimizes cumulative
+	 * error when applied repeatedly over a sequence of calculations.
+	 */
+	private static final Integer DEFAULT_PRECISION = MathContext.DECIMAL128.getPrecision();
+
+	private static final RoundingMode DEFAULT_PRECISION_ROUNDING_MODE = MathContext.DECIMAL128.getRoundingMode();
+
+	/**
+	 * Maximum integer value that can be accomodated as exponent in the
+	 * BigDecimal pow() function
+	 */
+	private static final int EXPONENT_MAX = 999999999;
+
+	// Internal fields
+
+	private BigDecimal value;
+
+	/**
+	 * MathContext of value.
+	 */
+	private MathContext precisionMathContext = new MathContext(DEFAULT_PRECISION, DEFAULT_PRECISION_ROUNDING_MODE);
+
+	private BigDecimal BIG_DECIMAL_ZERO = new BigDecimal(0, precisionMathContext);
+
+	private BigDecimal BIG_DECIMAL_ONE = new BigDecimal(1, precisionMathContext);
+
+	// Constructors
+
+	/**
+	 * Construct a BigDecimalProbabilityNumber from double type.
+	 * BigDecimal.valueOf() builds from a double type based on its canonical
+	 * representation. Precision of double type is at max 15.95 (~15).
+	 * 
+	 * @param value
+	 *            to be assigned.
+	 */
+	public BigDecimalProbabilityNumber(double value) {
+		if (value < 0 || value > 1) {
+			throw new IllegalArgumentException();
+		}
+		this.value = new BigDecimal(Double.toString(value), precisionMathContext);
+	}
+
+	/**
+	 * Construct a BigDecimalProbabilityNumber from BigDecimal type.
+	 * 
+	 * @param value
+	 *            to be assigned.
+	 */
+	public BigDecimalProbabilityNumber(BigDecimal value) {
+		if (null == value || value.compareTo(BIG_DECIMAL_ZERO) == -1 || value.compareTo(BIG_DECIMAL_ONE) == 1) {
+			throw new IllegalArgumentException();
+		}
+		this.value = new BigDecimal(value.toString(), precisionMathContext);
+	}
+
+	/**
+	 * Construct a BigDecimalProbabilityNumber from BigDecimal type.
+	 * 
+	 * @param value
+	 *            to be assigned.
+	 * 
+	 * @param mc
+	 *            MathContext to be associated with value.
+	 */
+	public BigDecimalProbabilityNumber(BigDecimal value, MathContext mc) {
+		precisionMathContext = mc;
+		setConstants();
+		if (value.compareTo(BIG_DECIMAL_ZERO) == -1 || value.compareTo(BIG_DECIMAL_ONE) == 1) {
+			throw new IllegalArgumentException();
+		}
```

Comment:
^ this is a common assertion check that could be moved into a private routine to help reduce duplicated code.